glean-core/src/event_database/mod.rs

```diff
@@ -100,26 +113,66 @@ impl EventDatabase {
     /// [Glean.initialize], but after we are ready to send pings, since this
-    /// could potentially collect and send pings.
+    /// could potentially collect and send the "events" ping.
     ///
     /// If there are any events queued on disk, it loads them into memory so
     /// that the memory and disk representations are in sync.
     ///
-    /// Secondly, if this is the first time the application has been run since
-    /// rebooting, any pings containing events are assembled into pings and cleared
-    /// immediately, since their timestamps won't be compatible with the timestamps
-    /// we would create during this boot of the device.
+    /// If event records for the "events" ping are present, they are assembled into
+    /// an "events" ping which is submitted immediately with reason "startup".
+    ///
+    /// If event records for custom pings are present, we increment the custom pings'
+    /// stores' `execution_counter` and record a `glean.restarted`
+    /// event with the current client clock in its `glean.startup.date` extra.
     ///
     /// # Arguments
     ///
     /// * `glean` - The Glean instance.
     ///
     /// # Returns
     ///
-    /// Whether at least one ping was generated.
+    /// Whether the "events" ping was submitted.
     pub fn flush_pending_events_on_startup(&self, glean: &Glean) -> bool {
         match self.load_events_from_disk() {
-            Ok(_) => self.send_all_events(glean),
+            Ok(_) => {
+                let stores_with_events: Vec<String> = {
+                    self.event_stores
+                        .read()
+                        .unwrap()
+                        .keys()
+                        .map(|x| x.to_owned())
+                        .collect() // safe unwrap, only error case is poisoning
+                };
+                // We do not want to be holding the event stores lock when
+                // submitting a ping or recording new events.
+                let has_events_events = stores_with_events.contains(&"events".to_owned());
+                let glean_restarted_stores = if has_events_events {
+                    stores_with_events
+                        .into_iter()
+                        .filter(|store| store != "events")
+                        .collect()
+                } else {
+                    stores_with_events
+                };
+                if !glean_restarted_stores.is_empty() {
+                    let glean_restarted = CommonMetricData {
+                        name: "restarted".into(),
+                        category: "glean".into(),
+                        send_in_pings: glean_restarted_stores,
+                        lifetime: Lifetime::Ping,
+                        ..Default::default()
+                    };
+                    let startup = get_iso_time_string(glean.start_time(), TimeUnit::Minute);
+                    let extra = [("glean.startup.date".into(), startup)].into();
```

Comment:
Is this why it's failing the schema checks on Android with Glean parser? ref: https://github.com/mozilla/glean_parser/blob/0960293637d40af31e3db356f91c8686a37801bf/glean_parser/schemas/metrics.2-0-0.schema.yaml#L28

`dotted snake case` looks like it may only expect a single dot? (or am I misreading the regex?)

This also begs the question, why is Kotlin the only language we are checking the schema in 🤔